controllers/metal3remediation_controller.go

```diff
@@ -0,0 +1,222 @@
+			remediationLog.Error(err, "unable to get metal3Remediation")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	helper, err := patch.NewHelper(metal3Remediation, r.Client)
+	if err != nil {
+		remediationLog.Error(err, "failed to init patch helper")
+		return ctrl.Result{}, errors.Wrap(err, "failed to init patch helper")
+	}
+	// Always patch metal3Remediation when exiting this function so we can persist any metal3Remediation changes.
+	defer func() {
+		err := helper.Patch(ctx, metal3Remediation)
+		if err != nil {
+			remediationLog.Error(err, "failed to Patch metal3Remediation")
+		}
+	}()
+
+	// Fetch the Machine.
+	capiMachine, err := util.GetOwnerMachine(ctx, r.Client, metal3Remediation.ObjectMeta)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "metal3Remediation's owner Machine could not be retrieved")
+	}
+	remediationLog = remediationLog.WithValues("unhealthy machine detected", capiMachine.Name)
+
+	// Fetch Metal3Machine
+	metal3Machine := capm3.Metal3Machine{}
+	key := client.ObjectKey{
+		Name:      capiMachine.Spec.InfrastructureRef.Name,
+		Namespace: capiMachine.Spec.InfrastructureRef.Namespace,
+	}
+	err = r.Get(ctx, key, &metal3Machine)
+	if err != nil {
+		remediationLog.Error(err, "metal3machine not found")
+		return ctrl.Result{}, errors.Wrapf(err, "metal3machine not found")
+	}
+
+	remediationLog = remediationLog.WithValues("metal3machine", metal3Machine.Name)
+
+	// Create a helper for managing the remediation object.
+	remediationMgr, err := r.ManagerFactory.NewRemediationManager(metal3Remediation, &metal3Machine, capiMachine, remediationLog)
+	if err != nil {
+		remediationLog.Error(err, "failed to create helper for managing the metal3remediation")
+		return ctrl.Result{}, errors.Wrapf(err, "failed to create helper for managing the metal3remediation")
+	}
+
+	// Handle deleted remediation
+	if !metal3Remediation.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, remediationMgr)
+	}
+
+	// Handle non-deleted remediation
+	return r.reconcileNormal(ctx, remediationMgr)
+}
+
+func (r *Metal3RemediationReconciler) reconcileNormal(ctx context.Context,
+	remediationMgr baremetal.RemediationManagerInterface,
+) (ctrl.Result, error) {
+
+	// If host is gone, exit early
+	host, _, err := remediationMgr.GetUnhealthyHost(ctx)
+	if err != nil {
+		return ctrl.Result{}, errors.Wrapf(err, "unable to find a host for unhealthy machine")
+	}
+
+	// If user has set bmh.Spec.Online to false
+	// do not try to remediate the host
+	if !remediationMgr.OnlineStatus(host) {
+		r.Log.Info("Host is powered off, unable to remediate")
+		remediationMgr.SetRemediationPhase(capm3.PhaseFailed)
+		return ctrl.Result{}, nil
+	}
+
+	remediationType := remediationMgr.GetRemediationType()
+
+	if remediationType != capm3.RebootRemediationStrategy {
+		r.Log.Info("unsupported remediation strategy")
+		return ctrl.Result{}, nil
+	}
+
+	if remediationType == capm3.RebootRemediationStrategy {
```

Comment:
I believe this line can be removed as we already checked this in line 134

Reply:
Should we keep this in place if we wanna support multiple remediation strategies in the future.

Reply:
seems like a reasonable idea, we already know there will be a new strategy that deletes Node objects